smartsim/_core/mli/infrastructure/storage/backbone_feature_store.py

```diff
@@ -352,68 +575,145 @@ def __init__(
-                    if not filter_set or msg.category in filter_set:
-                        messages.append(msg)
+        # Timeout elapsed or no messages received - return the empty list
+        if not all_message_bytes:
+            return events_received
 
-            # avoid getting stuck indefinitely waiting for the channel
-            elapsed = (time.time_ns() - start_at) / 1000000000
-            remaining = elapsed - self._global_timeout
-            if remaining > 0:
-                logger.debug(f"Consumer batch timeout exceeded by: {abs(remaining)}")
-                break
+        for message in all_message_bytes:
+            if not message or message is None:
+                continue
+
+            event = pickle.loads(message)
+            if not event:
+                logger.warning("Unable to unpickle message")
+
+            # skip events that don't pass a filter
+            if filter_set and event.category not in filter_set:
+                continue
+
+            events_received.append(event)
+
+        return events_received
+
+    def register(self) -> None:
+        """Send an event to register this consumer as a listener."""
+        descriptor = self._comm_channel.descriptor
+        event = OnCreateConsumer(descriptor, self._global_filters)
+
+        registrar_key = BackboneFeatureStore.MLI_BACKEND_CONSUMER
+        config = self._backbone.wait_for([registrar_key], 2.0)
+
+        registrar_descriptor = str(config.get(registrar_key, None))
+
+        if registrar_descriptor:
+            logger.debug(f"Sending registration for {self.name}")
+
+            registrar_channel = DragonCommChannel.from_descriptor(registrar_descriptor)
+            registrar_channel.send(bytes(event), timeout=1.0)
+
+            logger.debug(f"Registration for {self.name} sent")
+        else:
+            logger.warning("Unable to register. No registrar channel found.")
+
+    def listen_once(self, timeout: float = 0.001, batch_timeout: float = 1.0) -> None:
+        """Receives messages for the consumer a single time.
+
+        NOTE: Executes a single batch-retrieval to receive the maximum
+        number of messages available under batch timeout. To continually
+        listen, use `listen` in a non-blocking thread/process
+
+        :param timeout: Maximum time to wait (in seconds) for a message to arrive
+        :param timeout: Maximum time to wait (in seconds) for a batch to arrive
+        """
+        logger.debug(f"Starting event listener with {timeout} second timeout")
+        logger.debug("Awaiting new messages")
+
+        incoming_messages = self.recv(timeout=timeout, batch_timeout=batch_timeout)
+
+        if not incoming_messages:
+            logger.debug("Consumer received empty message list.")
 
-        return messages
+        for message in incoming_messages:
+            logger.debug(f"Sending event {message=} to handler.")
+            if self._event_handler:
```

Comment:
Maybe it would be worth adding a logger.debug() message if ``self._event_handler`` hasn't been set that messages received but no event handler has been set to process messages.

Also, for this class shoudl we try to work in the ``name`` attribute into logging in cases there are multiple consumers piped to the same output file?

Reply:
100$ on the debug log for a missing handler. We want to know if our consumer can't do what it's supposed to. Added

Reply:
Added names to log messages.